kubespawner/proxy.py

```diff
@@ -0,0 +1,210 @@
+import os
+import string
+import escapism
+import json
+from kubernetes import client
+
+from jupyterhub.proxy import Proxy
+from jupyterhub.utils import exponential_backoff
+
+from kubespawner.objects import make_ingress
+from kubespawner.reflector import NamespacedResourceReflector
+from concurrent.futures import ThreadPoolExecutor
+from traitlets import Unicode
+from tornado import gen
+from tornado.concurrent import run_on_executor
+
+
+class IngressReflector(NamespacedResourceReflector):
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_ingress'
+    api_group_name = 'ExtensionsV1beta1Api'
+
+    @property
+    def ingresses(self):
+        return self.resources
+
+class KubeIngressProxy(Proxy):
+    namespace = Unicode(
+        config=True,
+        help="""
+        Kubernetes namespace to spawn ingresses for single-user servers in.
+
+        If running inside a kubernetes cluster with service accounts enabled,
+        defaults to the current namespace. If not, defaults to 'default'
+        """
+    )
+
+    def _namespace_default(self):
+        """
+        Set namespace default to current namespace if running in a k8s cluster
+
+        If not in a k8s cluster with service accounts enabled, default to
+        'default'
+        """
+        ns_path = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+        if os.path.exists(ns_path):
+            with open(ns_path) as f:
+                return f.read().strip()
+        return 'default'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # other attributes
+        self.executor = ThreadPoolExecutor(max_workers=24)
```

Comment:
Are you sure you want to run this in 24 threads? It's probably appropriate to make this configurable if it should be any value other than one.

When we talk to the docker API, we use one thread, because it can't actually handle all those requests being concurrent anyway, so the key is that it's off the main thread, but the requests aren't really concurrent with each other.

Reply:
I've made this 3x concurrent_spawn_limit now